pkg/interactor/save_data_test.go

```diff
@@ -0,0 +1,113 @@
+package interactor
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	thingentities "github.com/CESARBR/knot-babeltower/pkg/thing/entities"
+	"github.com/CESARBR/knot-cloud-storage/pkg/entities"
+	"github.com/CESARBR/knot-cloud-storage/pkg/mocks"
+	"github.com/stretchr/testify/assert"
+)
+
+var date = time.Date(2020, time.April, 15, 0, 0, 0, 0, time.UTC)
+
+type SaveDataTestCase struct {
+	name               string
+	authorization      string
+	deviceID           string
+	payload            []entities.Payload
+	data               entities.Data
+	expectedData       *entities.Data
+	expectedDataError  error
+	expectedThing      []*thingentities.Thing
+	expectedThingError error
+	fakeLogger         *mocks.FakeLogger
+	fakeDataStore      *mocks.FakeDataStore
+	fakeService        *mocks.FakeService
+}
+
+var sdCases = []SaveDataTestCase{
+	{
+		"authorization key not provided",
+		"",
+		"",
+		nil,
+		entities.Data{},
+		nil,
+		nil,
+		nil,
+		errors.New("no access token provided"),
+		&mocks.FakeLogger{},
+		&mocks.FakeDataStore{},
+		&mocks.FakeService{},
+	},
+	{
+		"data successfully saved on the database",
+		"authorization-key",
+		"fc3fcf912d0c290a",
+		[]entities.Payload{{SensorID: 1, Value: 999}},
+		entities.Data{From: "fc3fcf912d0c290a", Payload: entities.Payload{SensorID: 1, Value: 999}, Timestamp: date},
+		nil,
+		nil,
+		nil,
+		nil,
+		&mocks.FakeLogger{},
+		&mocks.FakeDataStore{},
+		&mocks.FakeService{},
+	},
+	{
+		"failed to save the data on the database",
+		"authorization-key",
+		"fc3fcf912d0c290a",
+		[]entities.Payload{{SensorID: 1, Value: 999}},
+		entities.Data{From: "fc3fcf912d0c290a", Payload: entities.Payload{SensorID: 1, Value: 999}, Timestamp: date},
+		nil,
+		errors.New("error when trying to save data"),
+		[]*thingentities.Thing{
+			{
+				ID:    "fc3fcf912d0c290a",
+				Token: "token",
+				Name:  "thing",
+				Schema: []thingentities.Schema{
+					{
+						SensorID:  1,
+						ValueType: 3,
+						Unit:      0,
+						TypeID:    65521,
+						Name:      "Test",
+					},
+				},
+			},
+		},
+		nil,
+		&mocks.FakeLogger{},
+		&mocks.FakeDataStore{},
+		&mocks.FakeService{},
+	},
+}
+
+func TestSaveData(t *testing.T) {
+	for _, tc := range sdCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.fakeDataStore.
+				On("Save", tc.data).
```

Comment:
Shouldn't you pass `tc.payload` instead of `tc.data`?

Reply:
The DataStore Save function receives the entire data struct. If I change it to tc.payload, I would be passing to the database only the Payload object({ SensorID: ..., Value: ... }).

Reply:
Gotta.